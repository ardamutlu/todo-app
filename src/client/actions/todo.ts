export type Actions =
  | ReturnType<typeof getAllTodoRequest>
  | ReturnType<typeof getAllTodoSuccess>
  | ReturnType<typeof getAllTodoFailure>
  | ReturnType<typeof deleteTodoByIdRequest>
  | ReturnType<typeof deleteTodoByIdSuccess>
  | ReturnType<typeof deleteTodoByIdFailure>
  | ReturnType<typeof updateTodoStatusByIdRequest>
  | ReturnType<typeof updateTodoStatusByIdSuccess>
  | ReturnType<typeof updateTodoStatusByIdFailure>
  | ReturnType<typeof addNewTodoRequest>
  | ReturnType<typeof addNewTodoSuccess>
  | ReturnType<typeof addNewTodoFailure>
  | ReturnType<typeof updateTodoByIdRequest>
  | ReturnType<typeof updateTodoByIdSuccess>
  | ReturnType<typeof updateTodoByIdFailure>
  | ReturnType<typeof filterTodo>
  | ReturnType<typeof searchTodo>;

/**
 * GET ALL TODO
 */
export const GET_ALL_TODO_REQUEST = 'GET_ALL_TODO_REQUEST';
export const getAllTodoRequest = () =>
  ({
    type: GET_ALL_TODO_REQUEST
  } as const);

export const GET_ALL_TODO_SUCCESS = 'GET_ALL_TODO_SUCCESS';
export const getAllTodoSuccess = (payload: any) =>
  ({
    type: GET_ALL_TODO_SUCCESS,
    payload
  } as const);

export const GET_ALL_TODO_FAILURE = 'GET_ALL_TODO_FAILURE';
export const getAllTodoFailure = (err: Error) =>
  ({
    type: GET_ALL_TODO_FAILURE,
    payload: {
      err
    }
  } as const);

/**
 * DELETE TODO BY ID
 */
export const DELETE_TODO_BY_ID_REQUEST = 'DELETE_TODO_BY_ID_REQUEST';
export const deleteTodoByIdRequest = (payload: number) =>
  ({
    type: DELETE_TODO_BY_ID_REQUEST,
    payload
  } as const);

export const DELETE_TODO_BY_ID_SUCCESS = 'DELETE_TODO_BY_ID_SUCCESS';
export const deleteTodoByIdSuccess = (payload: any) =>
  ({
    type: DELETE_TODO_BY_ID_SUCCESS,
    payload
  } as const);

export const DELETE_TODO_BY_ID_FAILURE = 'DELETE_TODO_BY_ID_FAILURE';
export const deleteTodoByIdFailure = (err: Error) =>
  ({
    type: DELETE_TODO_BY_ID_FAILURE,
    payload: {
      err
    }
  } as const);

/**
 * UPDATE TODO STATUS BY ID
 */
export const UPDATE_TODO_STATUS_BY_ID_REQUEST = 'UPDATE_TODO_STATUS_BY_ID_REQUEST';
export const updateTodoStatusByIdRequest = (payload: any) =>
  ({
    type: UPDATE_TODO_STATUS_BY_ID_REQUEST,
    payload
  } as const);

export const UPDATE_TODO_STATUS_BY_ID_SUCCESS = 'UPDATE_TODO_STATUS_BY_ID_SUCCESS';
export const updateTodoStatusByIdSuccess = (payload: any) =>
  ({
    type: UPDATE_TODO_STATUS_BY_ID_SUCCESS,
    payload
  } as const);

export const UPDATE_TODO_STATUS_BY_ID_FAILURE = 'UPDATE_TODO_STATUS_BY_ID_FAILURE';
export const updateTodoStatusByIdFailure = (err: Error) =>
  ({
    type: UPDATE_TODO_STATUS_BY_ID_FAILURE,
    payload: {
      err
    }
  } as const);

/**
 * ADD NEW TODO
 */
export const ADD_NEW_TODO_REQUEST = 'ADD_NEW_TODO_REQUEST';
export const addNewTodoRequest = (payload: any) =>
  ({
    type: ADD_NEW_TODO_REQUEST,
    payload
  } as const);

export const ADD_NEW_TODO_SUCCESS = 'ADD_NEW_TODO_SUCCESS';
export const addNewTodoSuccess = (payload: any) =>
  ({
    type: ADD_NEW_TODO_SUCCESS,
    payload
  } as const);

export const ADD_NEW_TODO_FAILURE = 'ADD_NEW_TODO_FAILURE';
export const addNewTodoFailure = (err: Error) =>
  ({
    type: ADD_NEW_TODO_FAILURE,
    payload: {
      err
    }
  } as const);

/**
 * UPDATE TODO BY ID
 */
export const UPDATE_TODO_BY_ID_REQUEST = 'UPDATE_TODO_BY_ID_REQUEST';
export const updateTodoByIdRequest = (payload: any) =>
  ({
    type: UPDATE_TODO_BY_ID_REQUEST,
    payload
  } as const);

export const UPDATE_TODO_BY_ID_SUCCESS = 'UPDATE_TODO_BY_ID_SUCCESS';
export const updateTodoByIdSuccess = (payload: any) =>
  ({
    type: UPDATE_TODO_BY_ID_SUCCESS,
    payload
  } as const);

export const UPDATE_TODO_BY_ID_FAILURE = 'UPDATE_TODO_BY_ID_FAILURE';
export const updateTodoByIdFailure = (err: Error) =>
  ({
    type: UPDATE_TODO_BY_ID_FAILURE,
    payload: {
      err
    }
  } as const);

/**
 * FILTER TODO
 */
export const FILTER_TODO = 'FILTER_TODO';
export const filterTodo = (payload: any) =>
  ({
    type: FILTER_TODO,
    payload
  } as const);

/**
 * SEARCH TODO
 */
export const SEARCH_TODO = 'SEARCH_TODO';
export const searchTodo = (query: string) =>
  ({
    type: SEARCH_TODO,
    payload: {
      query
    }
  } as const);
