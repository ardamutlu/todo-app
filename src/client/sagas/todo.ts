import { END } from 'redux-saga';
import { put, call, takeLatest } from 'redux-saga/effects';
import {
  getAllTodoRequest,
  getAllTodoSuccess,
  getAllTodoFailure,
  deleteTodoByIdRequest,
  deleteTodoByIdSuccess,
  deleteTodoByIdFailure,
  updateTodoStatusByIdRequest,
  updateTodoStatusByIdSuccess,
  updateTodoStatusByIdFailure,
  updateTodoByIdRequest,
  updateTodoByIdSuccess,
  updateTodoByIdFailure,
  addNewTodoRequest,
  addNewTodoSuccess,
  addNewTodoFailure,
  GET_ALL_TODO_REQUEST,
  DELETE_TODO_BY_ID_REQUEST,
  UPDATE_TODO_STATUS_BY_ID_REQUEST,
  UPDATE_TODO_BY_ID_REQUEST,
  ADD_NEW_TODO_REQUEST
} from '../actions/todo';
import { fetchMethod } from './fetchMethod';
import { ErrorNotification } from '../templates';

function* getAllTodo(actions: ReturnType<typeof getAllTodoRequest>) {
  const options = { method: 'GET' };

  try {
    // async

    const body = yield fetchMethod({
      path: `/`,
      options
    });

    yield put(getAllTodoSuccess(body));
  } catch (err) {
    yield put(getAllTodoFailure(err));
  } finally {
    if (!process.env.IS_BROWSER) {
      yield call(stopSaga);
    }
  }
}

function* deleteTodoById(actions: ReturnType<typeof deleteTodoByIdRequest>) {
  const options = { method: 'DELETE' };

  try {
    // async

    const body = yield fetchMethod({
      path: `/${actions.payload}`,
      options
    });

    if (body.error) {
      yield call(ErrorNotification, 'error', body.message);
      yield put(deleteTodoByIdFailure(body));

      return;
    }

    yield put(deleteTodoByIdSuccess(body));
    yield call(ErrorNotification, 'success', 'Todo deleted successfully.');
  } catch (err) {
    yield put(deleteTodoByIdFailure(err));
  } finally {
    if (!process.env.IS_BROWSER) {
      yield call(stopSaga);
    }
  }
}

function* updateTodoStatusById(actions: ReturnType<typeof updateTodoStatusByIdRequest>) {
  const options = {
    method: 'PUT',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(
      Object.assign({}, actions.payload, { complete: !actions.payload.complete })
    )
  };

  try {
    // async

    const body = yield fetchMethod({
      path: `/${actions.payload.id}`,
      options
    });

    if (body.error) {
      yield call(ErrorNotification, 'error', body.message);
      yield put(updateTodoStatusByIdFailure(body));

      return;
    }

    yield put(updateTodoStatusByIdSuccess(body));
    yield call(ErrorNotification, 'success', 'Status changed successfully.');
  } catch (err) {
    yield put(updateTodoStatusByIdFailure(err));
  } finally {
    if (!process.env.IS_BROWSER) {
      yield call(stopSaga);
    }
  }
}

function* updateTodoById(actions: ReturnType<typeof updateTodoByIdRequest>) {
  const options = {
    method: 'PUT',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(actions.payload)
  };

  try {
    // async

    const body = yield fetchMethod({
      path: `/${actions.payload.id}`,
      options
    });

    if (body.error) {
      yield call(ErrorNotification, 'error', body.message);
      yield put(updateTodoByIdFailure(body));
      return;
    }

    yield put(updateTodoByIdSuccess(body));
    yield call(ErrorNotification, 'success', 'Todo updated successfully.');
  } catch (err) {
    yield put(updateTodoByIdFailure(err));
  } finally {
    if (!process.env.IS_BROWSER) {
      yield call(stopSaga);
    }
  }
}

function* addNewTodo(actions: ReturnType<typeof addNewTodoRequest>) {
  const options = {
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(actions.payload)
  };

  try {
    // async

    const body = yield fetchMethod({
      path: `/`,
      options
    });

    if (body.error) {
      yield call(ErrorNotification, 'error', body.message);
      yield put(addNewTodoFailure(body));
      return;
    }

    yield put(addNewTodoSuccess(body));
    yield call(ErrorNotification, 'success', 'Todo created successfully.');
  } catch (err) {
    yield put(addNewTodoFailure(err));
  } finally {
    if (!process.env.IS_BROWSER) {
      yield call(stopSaga);
    }
  }
}

function* stopSaga() {
  yield put(END);
}

export function* todoProcess() {
  yield takeLatest(GET_ALL_TODO_REQUEST, getAllTodo);
  yield takeLatest(DELETE_TODO_BY_ID_REQUEST, deleteTodoById);
  yield takeLatest(UPDATE_TODO_STATUS_BY_ID_REQUEST, updateTodoStatusById);
  yield takeLatest(UPDATE_TODO_BY_ID_REQUEST, updateTodoById);
  yield takeLatest(ADD_NEW_TODO_REQUEST, addNewTodo);
}
