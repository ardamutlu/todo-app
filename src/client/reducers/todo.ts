import * as actions from '../actions/todo';

export type State = {
  entities: Array<{}>;
  query: string;
  filter: {
    value: number;
    title: string;
  };
  loading: boolean;
  error: Error | null;
};

export const initialState: State = {
  entities: [],
  query: '',
  filter: {
    value: 0,
    title: 'All'
  },
  loading: false,
  error: null
};

export const reducer = (state: State = initialState, action: actions.Actions): State => {
  switch (action.type) {
    case actions.GET_ALL_TODO_REQUEST:
    case actions.DELETE_TODO_BY_ID_REQUEST:
    case actions.UPDATE_TODO_STATUS_BY_ID_REQUEST:
    case actions.UPDATE_TODO_BY_ID_REQUEST:
    case actions.ADD_NEW_TODO_REQUEST:
      return { ...state, loading: true };

    case actions.GET_ALL_TODO_SUCCESS:
    case actions.DELETE_TODO_BY_ID_SUCCESS:
    case actions.UPDATE_TODO_STATUS_BY_ID_SUCCESS:
    case actions.UPDATE_TODO_BY_ID_SUCCESS:
    case actions.ADD_NEW_TODO_SUCCESS:
      return { ...state, loading: false, entities: action.payload };

    case actions.GET_ALL_TODO_FAILURE:
    case actions.DELETE_TODO_BY_ID_FAILURE:
    case actions.UPDATE_TODO_STATUS_BY_ID_FAILURE:
    case actions.UPDATE_TODO_BY_ID_FAILURE:
    case actions.ADD_NEW_TODO_FAILURE:
      return { ...state, loading: false, error: action.payload.err };

    case actions.FILTER_TODO:
      return { ...state, filter: action.payload };

    case actions.SEARCH_TODO:
      return { ...state, query: action.payload.query };

    default:
      return state;
  }
};
