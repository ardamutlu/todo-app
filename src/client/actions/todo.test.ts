import * as actions from './todo';

test('should create an action to get all todo', () => {
  expect(actions.getAllTodoRequest()).toMatchSnapshot();
});

test('should create a success action to get all todo', () => {
  expect(actions.getAllTodoSuccess([])).toMatchSnapshot();
});

test('should create a failure action to get all', () => {
  expect(actions.getAllTodoFailure(new Error())).toMatchSnapshot();
});

test('should create an action to delete todo by id', () => {
  expect(actions.deleteTodoByIdRequest(0)).toMatchSnapshot();
});

test('should create a success action to delete todo by id', () => {
  expect(actions.deleteTodoByIdSuccess([])).toMatchSnapshot();
});

test('should create a failure action to delete todo by id', () => {
  expect(actions.deleteTodoByIdFailure(new Error())).toMatchSnapshot();
});

test('should create an action to update todo status by id', () => {
  expect(actions.updateTodoStatusByIdRequest({})).toMatchSnapshot();
});

test('should create a success action to update todo status by id', () => {
  expect(actions.updateTodoStatusByIdSuccess([])).toMatchSnapshot();
});

test('should create a failure action to update todo status by id', () => {
  expect(actions.updateTodoStatusByIdFailure(new Error())).toMatchSnapshot();
});

test('should create an action to add new todo', () => {
  expect(actions.addNewTodoRequest({})).toMatchSnapshot();
});

test('should create a success action to add new todo', () => {
  expect(actions.addNewTodoSuccess([])).toMatchSnapshot();
});

test('should create a failure action to add new todo', () => {
  expect(actions.addNewTodoFailure(new Error())).toMatchSnapshot();
});

test('should create an action to update todo by id', () => {
  expect(actions.updateTodoStatusByIdRequest({})).toMatchSnapshot();
});

test('should create a success action to update todo by id', () => {
  expect(actions.updateTodoStatusByIdSuccess([])).toMatchSnapshot();
});

test('should create a failure action to update todo by id', () => {
  expect(actions.updateTodoStatusByIdFailure(new Error())).toMatchSnapshot();
});

test('should create a action to filter todo', () => {
  expect(actions.filterTodo({})).toMatchSnapshot();
});

test('should create a action to search todo', () => {
  expect(actions.searchTodo('')).toMatchSnapshot();
});
