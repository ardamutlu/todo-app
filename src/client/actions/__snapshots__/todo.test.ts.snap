// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a action to filter todo 1`] = `
Object {
  "payload": Object {},
  "type": "FILTER_TODO",
}
`;

exports[`should create a action to search todo 1`] = `
Object {
  "payload": Object {
    "query": "",
  },
  "type": "SEARCH_TODO",
}
`;

exports[`should create a failure action to add new todo 1`] = `
Object {
  "payload": Object {
    "err": [Error],
  },
  "type": "ADD_NEW_TODO_FAILURE",
}
`;

exports[`should create a failure action to delete todo by id 1`] = `
Object {
  "payload": Object {
    "err": [Error],
  },
  "type": "DELETE_TODO_BY_ID_FAILURE",
}
`;

exports[`should create a failure action to get all 1`] = `
Object {
  "payload": Object {
    "err": [Error],
  },
  "type": "GET_ALL_TODO_FAILURE",
}
`;

exports[`should create a failure action to update todo by id 1`] = `
Object {
  "payload": Object {
    "err": [Error],
  },
  "type": "UPDATE_TODO_STATUS_BY_ID_FAILURE",
}
`;

exports[`should create a failure action to update todo status by id 1`] = `
Object {
  "payload": Object {
    "err": [Error],
  },
  "type": "UPDATE_TODO_STATUS_BY_ID_FAILURE",
}
`;

exports[`should create a success action to add new todo 1`] = `
Object {
  "payload": Array [],
  "type": "ADD_NEW_TODO_SUCCESS",
}
`;

exports[`should create a success action to delete todo by id 1`] = `
Object {
  "payload": Array [],
  "type": "DELETE_TODO_BY_ID_SUCCESS",
}
`;

exports[`should create a success action to get all todo 1`] = `
Object {
  "payload": Array [],
  "type": "GET_ALL_TODO_SUCCESS",
}
`;

exports[`should create a success action to update todo by id 1`] = `
Object {
  "payload": Array [],
  "type": "UPDATE_TODO_STATUS_BY_ID_SUCCESS",
}
`;

exports[`should create a success action to update todo status by id 1`] = `
Object {
  "payload": Array [],
  "type": "UPDATE_TODO_STATUS_BY_ID_SUCCESS",
}
`;

exports[`should create an action to add new todo 1`] = `
Object {
  "payload": Object {},
  "type": "ADD_NEW_TODO_REQUEST",
}
`;

exports[`should create an action to delete todo by id 1`] = `
Object {
  "payload": 0,
  "type": "DELETE_TODO_BY_ID_REQUEST",
}
`;

exports[`should create an action to get all todo 1`] = `
Object {
  "type": "GET_ALL_TODO_REQUEST",
}
`;

exports[`should create an action to update todo by id 1`] = `
Object {
  "payload": Object {},
  "type": "UPDATE_TODO_STATUS_BY_ID_REQUEST",
}
`;

exports[`should create an action to update todo status by id 1`] = `
Object {
  "payload": Object {},
  "type": "UPDATE_TODO_STATUS_BY_ID_REQUEST",
}
`;
